demos/http/http_demo_plaintext/http_demo_plaintext.c

```diff
@@ -0,0 +1,341 @@
+        }
+
+        if( pIndex == NULL )
+        {
+            /* Fail if no connection could be established. */
+            status = -1;
+        }
+        else
+        {
+            status = tcpSocket;
+        }
+    }
+
+    if( pListHead != NULL )
+    {
+        freeaddrinfo( pListHead );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief The transport send function provided to the MQTT context.
+ *
+ * @param[in] tcpSocket TCP socket.
+ * @param[in] pMessage Data to send.
+ * @param[in] bytesToSend Length of data to send.
+ *
+ * @return Number of bytes sent; negative value on error.
+ */
+static int32_t transportSend( int tcpSocket,
+                              const void * pMessage,
+                              size_t bytesToSend )
+{
+    return ( int32_t ) send( tcpSocket, pMessage, bytesToSend, 0 );
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief The transport receive function provided to the MQTT context.
+ *
+ * @param[in] tcpSocket TCP socket.
+ * @param[out] pBuffer Buffer for receiving data.
+ * @param[in] bytesToSend Size of pBuffer.
+ *
+ * @return Number of bytes received; negative value on error.
+ */
+static int32_t transportRecv( int tcpSocket,
+                              void * pBuffer,
+                              size_t bytesToRecv )
+{
+    return ( int32_t ) recv( tcpSocket, pBuffer, bytesToRecv, 0 );
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Send an HTTP request based on a specified method and path.
+ *
+ * @param[in] tcpSocket TCP socket.
+ * @param[in] pMethod The HTTP request method.
+ * @param[in] pPath The Request-URI to the objects of interest.
+ *
+ * @return #HTTP_SUCCESS if successful.
+ */
+static HTTPStatus_t _sendHttpRequest( int tcpSocket,
+                                      const char * pMethod,
+                                      const char * pPath )
+{
+    HTTPStatus_t returnStatus = HTTP_SUCCESS;
+    HTTPRequestHeaders_t requestHeaders = { 0 };
+    HTTPRequestInfo_t requestInfo = { 0 };
+    HTTPTransportInterface_t transport = { 0 };
+    HTTPResponse_t response = { 0 };
+
+    IotLog( "Sending HTTP %s request to %s%s\r\n",
+            pMethod, SERVER, pPath );
+
+    /* Initialize the request object. */
+    requestInfo.method = pMethod;
+    requestInfo.methodLen = sizeof( pMethod ) - 1;
+    requestInfo.pPath = pPath;
+    requestInfo.pathLen = sizeof( pPath ) - 1;
+
+    requestHeaders.pBuffer = userBuffer;
+    requestHeaders.bufferLen = USER_BUFFER_LENGTH;
+
+    returnStatus = HTTPClient_InitializeRequestHeaders( &requestHeaders,
+                                                        &requestInfo );
+
+    if( returnStatus == HTTP_SUCCESS )
+    {
+        /* Define the transport interface. */
+        transport.recv = transportRecv;
+        transport.send = transportSend;
+        transport.pContext = tcpSocket;
```

Comment:
This is giving me warnings because pContext is a struct while tcpSocket is an int. I think we need to avoid defining typedef HTTPNetworkContext_t in http_client.h and leave it up to application to define it. @sarenameas

Reply:
We need the context to be struct that holds both the socket and the socket error.